raft/src/main/java/org/apache/kafka/raft/internals/PartitionListener.java

```diff
@@ -0,0 +1,259 @@
+            return voterSetHistory.valueAtOrBefore(offset);
+        }
+    }
+
+    /**
+     * Returns the finalized kraft version at a given offset.
+     *
+     * @param offset the offset (inclusive)
+     * @return the finalized kraft version if one exist, otherwise 0
+     */
+    public short kraftVersionAtOffset(long offset) {
+        long fixedNextOffset = nextOffset;
+        if (offset >= fixedNextOffset) {
+            throw new IllegalArgumentException(
+                String.format(
+                    "Attempting the read the kraft.version at an offset (%d) which kraft hasn't seen (%d)",
+                    offset,
+                    fixedNextOffset - 1
+                )
+            );
+        }
+
+        synchronized (kraftVersionHistory) {
+            return kraftVersionHistory.valueAtOrBefore(offset).orElse((short) 0);
+        }
+    }
+
+    private void maybeLoadLog() {
+        while (log.endOffset().offset > nextOffset) {
+            LogFetchInfo info = log.read(nextOffset, Isolation.UNCOMMITTED);
+            try (RecordsIterator<?> iterator = new RecordsIterator<>(
+                    info.records,
+                    serde,
+                    bufferSupplier,
+                    maxBatchSizeBytes,
+                    true // Validate batch CRC
+                )
+            ) {
+                while (iterator.hasNext()) {
+                    Batch<?> batch = iterator.next();
+                    handleBatch(batch, OptionalLong.empty());
+                    nextOffset = batch.lastOffset() + 1;
+                }
+            }
+        }
+    }
+
+    private void maybeLoadSnapshot() {
+        if ((nextOffset == 0 || nextOffset < log.startOffset()) && log.latestSnapshot().isPresent()) {
+            RawSnapshotReader rawSnapshot = log.latestSnapshot().get();
+            // Clear the current state
+            synchronized (kraftVersionHistory) {
+                kraftVersionHistory.clear();
+            }
+            synchronized (voterSetHistory) {
+                voterSetHistory.clear();
+            }
+
+            // Load the snapshot since the listener is at the start of the log or the log doesn't have the next entry.
+            try (SnapshotReader<?> reader = RecordsSnapshotReader.of(
+                    rawSnapshot,
+                    serde,
+                    bufferSupplier,
+                    maxBatchSizeBytes,
+                    true // Validate batch CRC
+                )
+            ) {
+                logger.info(
+                    "Loading snapshot ({}) since log start offset ({}) is greater than the internal listener's next offset ({})",
+                    reader.snapshotId(),
+                    log.startOffset(),
+                    nextOffset
+                );
+                OptionalLong currentOffset = OptionalLong.of(reader.lastContainedLogOffset());
+                while (reader.hasNext()) {
+                    Batch<?> batch = reader.next();
```

Comment:
Slightly annoying that we load the full batch into memory. It ought to be possible to look at the header only to see if it is a control batch and skip loading otherwise. Potential optimization for the future I guess.

Reply:
I agree. I created [KAFKA-16517](https://issues.apache.org/jira/browse/KAFKA-16517) to fix this later.